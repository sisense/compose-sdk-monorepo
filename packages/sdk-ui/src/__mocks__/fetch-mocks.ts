import { MockParams } from 'vitest-fetch-mock';
import mockGlobals from './data/mock-globals.json';
import mockPalettes from './data/mock-palettes.json';
import mockWidgetDrilldown from './data/mock-widget-drilldown.json';
import mockJaqlDrilldown from './data/mock-jaql-drilldown.json';
import mockJaqlDates from './data/mock-jaql-dates.json';

export const mockUrl = `http://10.50.0.1:30845`;
export const mockToken = 'fake-token';
export const mockDashboardId = 'fake-dashboard-id';
export const mockWidgetId = 'fake-widget-id';

export const fetchMocks: Record<string, [string, MockParams]> = {
  globals: [JSON.stringify(mockGlobals), { status: 200 }],
  palettes: [JSON.stringify(mockPalettes), { status: 200 }],
  widgetDrilldown: [JSON.stringify(mockWidgetDrilldown), { status: 200 }],
  jaqlDrilldown: [JSON.stringify(mockJaqlDrilldown), { status: 200 }],
  jaqlDates: [JSON.stringify(mockJaqlDates), { status: 200 }],
};
