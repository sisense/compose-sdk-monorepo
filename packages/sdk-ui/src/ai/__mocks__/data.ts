import { DataModel, Perspective } from '@/ai/api/types';

export const dataModels: DataModel[] = [
  {
    oid: 'm1',
    title: 'Model 1',
  },
  {
    oid: 'm2',
    title: 'Model 2',
  },
  {
    oid: 'm3',
    title: 'Model 3',
  },
];

export const perspectives: Perspective[] = [
  {
    oid: 'p1',
    name: 'Perspective 1',
    description: 'This is my first perspective',
    isDefault: false,
  },
  {
    oid: 'p2',
    name: 'Perspective 2',
    description: 'This is my second perspective',
    isDefault: true,
  },
  {
    oid: 'p3',
    name: 'Perspective 3',
    description: 'This is my third perspective',
    isDefault: false,
  },
];
