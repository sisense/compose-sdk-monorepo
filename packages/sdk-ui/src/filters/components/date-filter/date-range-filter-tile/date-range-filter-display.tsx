import React, { useMemo } from 'react';
import { DateRangeFilter } from '@sisense/sdk-data';
import { useTranslation } from 'react-i18next';
import { FilterContentDisplay } from '@/filters/components/common/filter-content-display';

type DateRangeFilterDisplayProps = {
  filter: DateRangeFilter;
};

type DisplayVariant = 'from-to' | 'only-from' | 'only-to';

export const DateRangeFilterDisplay: React.FC<DateRangeFilterDisplayProps> = ({ filter }) => {
  const { t } = useTranslation();
  const { from, to } = filter;

  const displayVariant: DisplayVariant = from && to ? 'from-to' : from ? 'only-from' : 'only-to';

  const textContent = useMemo(() => {
    switch (displayVariant) {
      case 'from-to':
        return t('dateFilter.dateRange.fromTo', { from, to });
      case 'only-from':
        return t('dateFilter.dateRange.from', { val: from });
      case 'only-to':
        return t('dateFilter.dateRange.to', { val: to });
      default:
        return '';
    }
  }, [displayVariant, from, to, t]);

  return <FilterContentDisplay>{textContent}</FilterContentDisplay>;
};
