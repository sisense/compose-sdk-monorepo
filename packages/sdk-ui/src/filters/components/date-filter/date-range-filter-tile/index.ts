export * from './date-range-filter-tile';
