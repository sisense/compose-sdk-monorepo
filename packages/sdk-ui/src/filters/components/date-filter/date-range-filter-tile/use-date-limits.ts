import { DataSource, LevelAttribute, measures } from '@sisense/sdk-data';
import { useSisenseContext } from '../../../../sisense-context/sisense-context';
import { useEffect, useState } from 'react';
import { executeQuery } from '../../../../query/execute-query';

type DateLimits = {
  minDate?: string;
  maxDate?: string;
};

export const useDateLimits = (
  userLimits: DateLimits,
  attribute: LevelAttribute,
  dataSource?: DataSource,
): Required<DateLimits> | null => {
  const [fetchedLimits, setFetchedLimits] = useState<Required<DateLimits> | null>(null);
  const { isInitialized: isAppInitialized, app } = useSisenseContext();
  const shouldFetchLimits = !isLimitsDefinedByUser(userLimits) && isAppInitialized && app;
  useEffect(() => {
    if (!shouldFetchLimits) {
      return;
    }

    void executeQuery(
      dataSource,
      [],
      [measures.min(attribute), measures.max(attribute)],
      [],
      [],
      app,
    ).then((data) => {
      const queryMembers = data.rows[0];
      setFetchedLimits({
        minDate: queryMembers[0].data as string,
        maxDate: queryMembers[1].data as string,
      });
    });
  }, [app, attribute, dataSource, shouldFetchLimits]);
  return mergeDateLimits(userLimits, fetchedLimits);
};

function isLimitsDefinedByUser(userLimits: DateLimits): userLimits is Required<DateLimits> {
  return userLimits.minDate !== undefined && userLimits.maxDate !== undefined;
}

function mergeDateLimits(
  userLimits: DateLimits,
  serverLimits: Required<DateLimits> | null,
): Required<DateLimits> | null {
  const minDate = userLimits.minDate ?? serverLimits?.minDate;
  const maxDate = userLimits.maxDate ?? serverLimits?.maxDate;
  return minDate && maxDate ? { minDate, maxDate } : null;
}
