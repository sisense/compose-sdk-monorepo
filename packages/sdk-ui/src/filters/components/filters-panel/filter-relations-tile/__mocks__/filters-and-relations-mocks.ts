import { FilterRelationsRules } from '@/utils/filter-relations';
import { Filter, filterFactory } from '@ethings-os/sdk-data';
import * as DM from '@/__test-helpers__/sample-ecommerce';

const countryFilter = filterFactory.members(DM.Country.Country, []);
const yearsFilter = filterFactory.members(DM.Commerce.Date.Years, []);
const revenueFilter = filterFactory.greaterThan(DM.Commerce.Revenue, 0);
const conditionFilter = filterFactory.members(DM.Commerce.Condition, ['Unspecified']);
const genderFilter = filterFactory.members(DM.Commerce.Gender, []);

export const filtersMock: Filter[] = [
  countryFilter,
  yearsFilter,
  revenueFilter,
  conditionFilter,
  genderFilter,
];

export const relationsMock: FilterRelationsRules = {
  left: {
    left: {
      instanceid: countryFilter.config.guid,
    },
    right: {
      instanceid: yearsFilter.config.guid,
    },
    operator: 'OR',
  },
  right: {
    left: {
      instanceid: revenueFilter.config.guid,
    },
    right: {
      left: {
        instanceid: conditionFilter.config.guid,
      },
      right: {
        instanceid: genderFilter.config.guid,
      },
      operator: 'AND',
    },
    operator: 'OR',
  },
  operator: 'AND',
};
