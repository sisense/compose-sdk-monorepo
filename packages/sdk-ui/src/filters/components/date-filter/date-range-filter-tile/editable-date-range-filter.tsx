import { DateRangeFilter, filterFactory } from '@sisense/sdk-data';

import { DateFilter } from '../date-filter';
import type { DateRangeFilterTileProps } from './date-range-filter-tile';
import { DateLimits } from './use-date-limits';

type EditableDateRangeFilterProps = DateRangeFilterTileProps & {
  filter: DateRangeFilter;
  dateLimits: Required<DateLimits>;
  isOldDateRangeFilterTile?: boolean;
  disabled?: boolean;
};

/**
 * Editable Date Range Filter component for filtering data by date range.
 */
export const EditableDateRangeFilter = (props: EditableDateRangeFilterProps) => {
  const {
    filter,
    attribute,
    onChange,
    parentFilters: rawParentFilters,
    dateLimits,
    isOldDateRangeFilterTile,
    disabled,
  } = props;
  const parentFilters =
    rawParentFilters && rawParentFilters?.filter((filter) => filter && filter !== null);

  const fromDate = filter.from || dateLimits?.minDate;
  const toDate = filter.to || dateLimits?.maxDate;

  return (
    <DateFilter
      onChange={(dateFilter) => {
        const newFilter = filterFactory.dateRange(
          attribute,
          dateFilter.filter.from,
          dateFilter.filter.to,
        );
        onChange(newFilter);
      }}
      value={{
        from: fromDate,
        to: toDate,
      }}
      limit={dateLimits}
      isDependent={parentFilters && parentFilters.length > 0}
      variant={isOldDateRangeFilterTile ? 'grey' : 'white'}
      disabled={disabled}
    />
  );
};
