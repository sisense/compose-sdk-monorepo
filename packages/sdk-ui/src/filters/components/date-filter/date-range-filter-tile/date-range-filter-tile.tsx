import {
  LevelAttribute,
  DataSource,
  Filter,
  DateRangeFilter,
  FilterTypes,
} from '@sisense/sdk-data';
import cloneDeep from 'lodash/cloneDeep';
import { useState } from 'react';
import { asSisenseComponent } from '../../../../decorators/component-decorators/as-sisense-component';
import { FilterTile } from '../../filter-tile';
import { EditableDateRangeFilter } from './editable-date-range-filter';
import { DateRangeFilterDisplay } from './date-range-filter-display';
import { TranslatableError } from '@/translation/translatable-error';
import { AnyObject } from '@/utils/utility-types';
import { useDateLimits } from './use-date-limits';

export interface DateRangeFilterTileProps {
  /**
   * Filter tile title
   */
  title: string;
  /**
   * Date level attribute the filter is based on
   */
  attribute: LevelAttribute;
  /**
   * Data source the query is run against - e.g. `Sample ECommerce`
   *
   * If not specified, the query will use the `defaultDataSource` specified in the parent Sisense Context.
   */
  dataSource?: DataSource;
  /**
   * Date range filter.
   */
  filter: Filter;
  /**
   * Earliest allowed date for selection.
   *
   * If not specified, the earliest date of the target date-level attribute will be used.
   */
  earliestDate?: string;
  /**
   * Latest allowed date for selection.
   *
   * If not specified, the latest date of the target date-level attribute will be used.
   */
  lastDate?: string;
  /**
   * Callback function that is called when the date range filter object should be updated.
   *
   * @param filter - Date range filter
   */
  onChange: (filter: Filter) => void;
  /**
   * List of filters this filter is dependent on.
   */
  parentFilters?: Filter[];

  /**
   * Whether to display the filter as a tiled version.
   * @default false
   * @internal
   */
  tiled?: boolean;
}

/**
 * Date Range Filter Tile component for filtering data by date range.
 *
 * @example
 * React example of configuring the date min max values and handling onChange event.
 * ```tsx
 * const [dateRangeFilter, setDateRangeFilter] = useState<Filter>(
 *   filterFactory.dateRange(DM.Commerce.Date.Years),
 * );
 *
 * return (
 *   <DateRangeFilterTile
 *     title="Date Range"
 *     attribute={DM.Commerce.Date.Years}
 *     filter={dateRangeFilter}
 *     onChange={(filter: Filter) => {
 *       setDateRangeFilter(filter);
 *     }}
 *   />
 * );
 * ```
 *
 * <img src="media://date-filter-example-1.png" width="800px" />
 * @param props - Date Range Filter Tile Props
 * @returns Date Range Filter Tile component
 * @group Filter Tiles
 */
export const DateRangeFilterTile = asSisenseComponent({ componentName: 'DateRangeFilterTile' })(
  (props: DateRangeFilterTileProps) => {
    const {
      filter,
      onChange,
      title,
      earliestDate,
      lastDate,
      attribute,
      dataSource,
      parentFilters,
      tiled = false,
    } = props;
    if (!isDateRangeFilter(filter)) {
      throw new TranslatableError('errors.invalidFilterType');
    }
    const [lastFilter, setLastFilter] = useState(filter);
    let disabled = lastFilter.disabled;

    const onUpdateValues = (newFilter: DateRangeFilter) => {
      if (newFilter) {
        setLastFilter(newFilter);
      }
      onChange(newFilter);
    };

    const dateLimits = useDateLimits(
      {
        minDate: earliestDate,
        maxDate: lastDate,
      },
      attribute,
      dataSource,
      parentFilters,
    );

    if (!tiled) {
      if (!dateLimits) {
        return null;
      }
      return (
        <EditableDateRangeFilter
          {...props}
          filter={filter}
          dateLimits={dateLimits}
          isOldDateRangeFilterTile
        />
      );
    }

    return (
      <FilterTile
        title={title}
        renderContent={(collapsed) => {
          return collapsed || !dateLimits ? (
            <DateRangeFilterDisplay filter={filter} />
          ) : (
            <div className="csdk-mt-2 csdk-ml-2 csdk-mb-1">
              <EditableDateRangeFilter
                {...props}
                filter={filter}
                dateLimits={dateLimits}
                disabled={disabled}
              />
            </div>
          );
        }}
        onToggleDisabled={() => {
          disabled = !disabled;
          const newFilter = cloneDeep(lastFilter);
          newFilter.disabled = disabled;
          onUpdateValues(newFilter);
        }}
        disabled={disabled}
      />
    );
  },
);

function isDateRangeFilter(filter: Filter & AnyObject): filter is DateRangeFilter {
  return 'filterType' in filter && filter.filterType === FilterTypes.date;
}
