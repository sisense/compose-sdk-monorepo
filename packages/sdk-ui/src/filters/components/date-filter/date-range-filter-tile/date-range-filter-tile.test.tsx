/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { mockUrl, mockToken, fetchMocks } from '../../../../__mocks__/fetch-mocks';

import { fireEvent, render, screen } from '@testing-library/react';
import { DateRangeFilterTile } from './date-range-filter-tile';
import * as DM from '../../../../__mocks__/data/sample-ecommerce-autogenerated';
import { filterFactory } from '@sisense/sdk-data';
import { SisenseContextProvider } from '../../../../sisense-context/sisense-context-provider';

describe('DateRangeFilterTile', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it('should render DateRangeFilterTile', async () => {
    expect.assertions(2);

    // Rendering a DateRangeFilterTile requires 3 fetches
    fetchMock.mockResponses(fetchMocks.globals, fetchMocks.palettes, fetchMocks.jaqlDates);

    const dateRangeFilter = filterFactory.dateRange(DM.Commerce.Date.Years);

    const { container } = render(
      <SisenseContextProvider url={mockUrl} token={mockToken} enableTracking={false}>
        <DateRangeFilterTile
          title="Date Range"
          dataSource={DM.DataSource}
          attribute={DM.Commerce.Date.Years}
          filter={dateRangeFilter}
          onChange={() => {}}
        />
      </SisenseContextProvider>,
    );

    expect(await screen.findByText('2009-01-01')).toBeInTheDocument();

    const dateButton = container.querySelector('.h-button') || container;
    fireEvent.click(dateButton);
    expect(dateButton).toMatchSnapshot();
  });

  it('should render a DateRangeFilterTile component with jaql error', async () => {
    expect.assertions(1);

    // Set failed jaql response
    fetchMock.mockResponses(fetchMocks.globals, fetchMocks.palettes, [
      JSON.stringify({ error: true }),
      { status: 500 },
    ]);

    const dateRangeFilter = filterFactory.dateRange(DM.Commerce.Date.Years);

    const { findByLabelText, findByText } = render(
      <SisenseContextProvider url={mockUrl} token={mockToken} enableTracking={false}>
        <DateRangeFilterTile
          title="Date Range"
          dataSource={DM.DataSource}
          attribute={DM.Commerce.Date.Years}
          filter={dateRangeFilter}
          onChange={() => {}}
        />
      </SisenseContextProvider>,
    );

    const errorBoxContainer = await findByLabelText('error-box');
    fireEvent.mouseEnter(errorBoxContainer);
    const errorBoxText = await findByText(/Error/);

    expect(errorBoxText).toBeTruthy();
  });
});
