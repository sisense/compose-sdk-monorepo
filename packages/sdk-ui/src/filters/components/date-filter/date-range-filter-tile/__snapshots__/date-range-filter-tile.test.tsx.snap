// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateRangeFilterTile > should render DateRangeFilterTile 1`] = `
<div>
  <div>
    <div
      aria-label="date range filter"
      class="csdk-w-fit csdk-flex csdk-flex-wrap"
    >
      <div
        aria-label="DateRangeField"
        class="csdk-relative csdk-flex csdk-mr-[5px] csdk-cursor-pointer"
      >
        <label
          class="csdk-text-text-content csdk-my-[5px] csdk-mr-[7px] csdk-text-[13px] csdk-leading-[18px] css-1ny4au0"
        >
          From
        </label>
        <div
          class="csdk-relative csdk-h-6"
        >
          <button
            aria-label="DateRangeFieldButton"
            class="csdk-text-left csdk-w-[152px] csdk-bg-[#f4f4f8] csdk-text-[13px] csdk-outline-0 csdk-border csdk-border-transparent csdk-p-input csdk-h-6 csdk-rounded-[4px] disabled:csdk-placeholder:csdk-opacity-30 disabled:csdk-cursor-not-allowed focus:csdk-border-solid focus:csdk-border-input focus:csdk-border-UI-default hover:csdk-border-guiding csdk-text-text-active csdk-pl-2.5 csdk-pr-10 csdk-overflow-hidden  css-1yw6u95"
            label="From"
            value="2009-01-01"
          >
            2009-01-01
            <div
              aria-label="DateRangeFieldIcon"
              class="csdk-absolute csdk-right-[4px] csdk-top-1/2 csdk-transform -csdk-translate-y-1/2"
            >
              <svg
                class="csdk-flex csdk-text-text-active"
                fill="none"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9 5V3.5C9 3.22386 9.22386 3 9.5 3C9.77614 3 10 3.22386 10 3.5V5H13V3.5C13 3.22386 13.2239 3 13.5 3C13.7761 3 14 3.22386 14 3.5V5H17C18.1046 5 19 5.89543 19 7V8V9V17C19 18.1046 18.1046 19 17 19H6C4.89543 19 4 18.1046 4 17V9V8V7C4 5.89543 4.89543 5 6 5H9ZM10 6H13V6.5C13 6.77614 13.2239 7 13.5 7C13.7761 7 14 6.77614 14 6.5V6H17C17.5523 6 18 6.44772 18 7V8H5V7C5 6.44772 5.44772 6 6 6H9V6.5C9 6.77614 9.22386 7 9.5 7C9.77614 7 10 6.77614 10 6.5V6ZM5 17V9H18V17C18 17.5523 17.5523 18 17 18H6C5.44772 18 5 17.5523 5 17ZM14.5 12H8.5C8.22386 12 8 11.7761 8 11.5C8 11.2239 8.22386 11 8.5 11H14.5C14.7761 11 15 11.2239 15 11.5C15 11.7761 14.7761 12 14.5 12ZM8.5 14H14.5C14.7761 14 15 13.7761 15 13.5C15 13.2239 14.7761 13 14.5 13H8.5C8.22386 13 8 13.2239 8 13.5C8 13.7761 8.22386 14 8.5 14ZM14.5 16H8.5C8.22386 16 8 15.7761 8 15.5C8 15.2239 8.22386 15 8.5 15H14.5C14.7761 15 15 15.2239 15 15.5C15 15.7761 14.7761 16 14.5 16Z"
                  fill="#5B6372"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
      <div
        aria-label="DateRangeField"
        class="csdk-relative csdk-flex csdk-mr-[5px] csdk-cursor-pointer"
      >
        <label
          class="csdk-text-text-content csdk-my-[5px] csdk-mr-[7px] csdk-text-[13px] csdk-leading-[18px] css-1ny4au0"
        >
          To
        </label>
        <div
          class="csdk-relative csdk-h-6"
        >
          <button
            aria-label="DateRangeFieldButton"
            class="csdk-text-left csdk-w-[152px] csdk-bg-[#f4f4f8] csdk-text-[13px] csdk-outline-0 csdk-border csdk-border-transparent csdk-p-input csdk-h-6 csdk-rounded-[4px] disabled:csdk-placeholder:csdk-opacity-30 disabled:csdk-cursor-not-allowed focus:csdk-border-solid focus:csdk-border-input focus:csdk-border-UI-default hover:csdk-border-guiding csdk-text-text-active csdk-pl-2.5 csdk-pr-10 csdk-overflow-hidden  css-1yw6u95"
            label="To"
            value="2013-01-01"
          >
            2013-01-01
            <div
              aria-label="DateRangeFieldIcon"
              class="csdk-absolute csdk-right-[4px] csdk-top-1/2 csdk-transform -csdk-translate-y-1/2"
            >
              <svg
                class="csdk-flex csdk-text-text-active"
                fill="none"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9 5V3.5C9 3.22386 9.22386 3 9.5 3C9.77614 3 10 3.22386 10 3.5V5H13V3.5C13 3.22386 13.2239 3 13.5 3C13.7761 3 14 3.22386 14 3.5V5H17C18.1046 5 19 5.89543 19 7V8V9V17C19 18.1046 18.1046 19 17 19H6C4.89543 19 4 18.1046 4 17V9V8V7C4 5.89543 4.89543 5 6 5H9ZM10 6H13V6.5C13 6.77614 13.2239 7 13.5 7C13.7761 7 14 6.77614 14 6.5V6H17C17.5523 6 18 6.44772 18 7V8H5V7C5 6.44772 5.44772 6 6 6H9V6.5C9 6.77614 9.22386 7 9.5 7C9.77614 7 10 6.77614 10 6.5V6ZM5 17V9H18V17C18 17.5523 17.5523 18 17 18H6C5.44772 18 5 17.5523 5 17ZM14.5 12H8.5C8.22386 12 8 11.7761 8 11.5C8 11.2239 8.22386 11 8.5 11H14.5C14.7761 11 15 11.2239 15 11.5C15 11.7761 14.7761 12 14.5 12ZM8.5 14H14.5C14.7761 14 15 13.7761 15 13.5C15 13.2239 14.7761 13 14.5 13H8.5C8.22386 13 8 13.2239 8 13.5C8 13.7761 8.22386 14 8.5 14ZM14.5 16H8.5C8.22386 16 8 15.7761 8 15.5C8 15.2239 8.22386 15 8.5 15H14.5C14.7761 15 15 15.2239 15 15.5C15 15.7761 14.7761 16 14.5 16Z"
                  fill="#5B6372"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
